### ISSUE TITLE
Ignore front matter in Markdown

### ISSUE THREAD

**Issue by Cryolitia** [translate/translate#5399]
### Describe the problem

Nowadays, many blog frame would use markdown with front matter, it's a small piece at the start of a markdown file, it looks like:

```markdown
---
date: 2024-02-02T04:14:54-08:00
draft: false
params:
  author: John Smith
title: Example
weight: 10
---
```

or in hugo may be

```markdown
+++
date = 2024-02-02T04:14:54-08:00
draft = false
title = 'Example'
weight = 10
[params]
  author = 'John Smith'
+++
```

Now, weblate will only parse this picec as plain text like `title: Base Pitch parent: Pitch Naturalisation nav_order: 2`

I would like to let translate parse or just ignore it.

hugo: https://gohugo.io/content-management/front-matter/
vitepress: https://vitepress.dev/zh/guide/frontmatter

**Comment by niccokunzmann** [comment]
yaml is already supported.

I would guess, one can pass the lines over to the yaml parser:

https://github.com/translate/translate/blob/e7753def917ea341795d4f2db025a3f8d8b8f45c/translate/storage/markdown.py#L97

This PR will help show where and how to integrate it:

https://github.com/translate/translate/pull/4966

Possibilities I see:

1. The markdown **parser** can hand parsing over to yaml
2. The conversion tool checks for yaml between `---` at the start of the file and uses that parser, too.

I think, option 2 is cool because markdown does not always imply that YAML can be is present. That is the case for files, not e.g. for strings of documents that are in markdown. e.g. a multiline form for a website. Just because it is split up, it does not mean that there is YAML always possible at the front.

**Comment by stevepiercy** [comment]
In a few Python projects that use [MyST](https://myst-parser.readthedocs.io/en/latest/), here's another example of front matter.

```md
---
myst:
  html_meta:
    "description": "This website is the documentation of Plone, an open source, enterprise-level content management system."
    "property=og:description": "This website is the documentation of Plone, an open source, enterprise-level content management system."
    "property=og:title": "Plone 6 Documentation"
    "keywords": "Plone 6, content management system, CMS, open source, Documentation, Volto, Classic UI, frontend, backend, plone.restapi, plone.api"
---
```

Here I use a Sphinx extension to parse open graph values to improve SEO indexing for English. I think allowing the values to be translated would be very helpful.

**Comment by niccokunzmann** [comment]
GitHub describes it here: https://docs.github.com/en/contributing/writing-for-github-docs/using-yaml-frontmatter
But to me, it appears that one can have any yaml in it. So it is unclear which values actually need translating.
There seem to be some that are used by others, too that are useful inside to translate:

|key | [jekyll](https://docs.github.com/en/contributing/writing-for-github-docs/using-yaml-frontmatter) | [hugo](https://gohugobrasil.netlify.app/content-management/front-matter/) | [MyST](https://myst-parser.readthedocs.io/en/latest/configuration.html#frontmatter-local-configuration) in `myst.html_meta` |
| ---- | ----  | ---- | ---- |
| title |  YES  | YES   | ?   |
| shortTitle | YES   |    NO|  NO  |
| intro |  YES  |  NO  |  NO  |
| topics | YES   |  NO  |  NO  |
| description | NO | YES | YES   |
| keywords  | NO | YES |  YES  |
| linkTitle | NO | YES |  NO  |
| categories  | NO | YES | NO   |

**Comment by stevepiercy** [comment]
Here's the MyST reference for frontmatter.

https://mystmd.org/guide/frontmatter

Per https://myst-parser.readthedocs.io/en/latest/syntax/reference.html#extended-block-tokens, frontmatter is placed at the start of a document and consists of a YAML block enclosed by `---` on separate lines.

In Pelican, Sphinx, and docutils—all of which use the [reStructuredText specification](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#field-lists)—use [field lists](https://www.sphinx-doc.org/en/master/usage/restructuredtext/field-lists.html) to perform a similar frontmatter function. I think, but I'm not certain, that MyST passes configuration keys to docutils.

This utility may be a useful example to parse some YAML keys as configuration and others to render as HTML.

https://myst-parser.readthedocs.io/en/latest/_modules/myst_parser/mdit_to_docutils/base.html#DocutilsRenderer.render_front_matter

@niccokunzmann there are extensions in MyST that provide support for additional frontmatter, including any Open Graph data. And you don't need to put the values in `myst.html_meta` but use native Sphinx frontmatter support through field lists. So that column for MyST in the table you cite is somewhat misleading. See https://mystmd.org/guide/frontmatter#available-frontmatter-fields for what is actually supported.

**Comment by nijel** [comment]
I don't think we should go into parsing the front matter content. https://github.com/translate/translate/pull/5659 will make it parse it as a header and keep it intact in round trip.
